src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,453 @@
+
+Here are the steps to take before migrating this package to sound null safety:
+
+1. Wait for `examples_util` to be migrated.
+
+2. Update `pubspec.yaml` to use the latest compatible releases of
+   `pedantic`, `test`, and `examples_util`
+   (as listed in the **Resolvable** column).
+   For example, the new dev dependencies might look like this:
+   ```dart
+dev_dependencies:
+  examples_util: ^0.0.3-nullsafety.0
+  pedantic: ^1.10.0-nullsafety.3
+  test: ^1.16.0-nullsafety.9
+```
+
+3. Run `dart pub upgrade`.
+
+4. Run `dart pub outdated --mode=null-safety` again,
+   to make sure that all dependencies are null safe and up-to-date.
+
+
+## 2. Migrate {#step2-migrate}
+
+Most of the changes that your code needs to be null safe
+are easily predictable.
+For example, if a variable can be `null`,
+its type needs a `?` suffix.
+A named parameter that shouldn't be nullable
+needs to be marked `required`.
+
+The migration tool can make most of these changes,
+and you can add hints to customize the changes.
+Alternatively, you can
+[migrate your code by hand](#migrating-by-hand).
+
+{{ site.alert.tip }}
+  For additional help while migrating code, check the
+  [null safety FAQ][].
+{{ site.alert.end }}
+
+[null safety FAQ]: /null-safety/faq
+
+
+### Using the migration tool
+
+The migration tool takes a package of null-unsafe Dart code
+and converts it to null safety.
+Before starting the tool, make sure you're ready:
+
+* Use the latest 2.12 beta release of the Dart SDK.
+* Use `dart pub outdated --mode=null-safety` to make sure that
+  all dependencies are null safe and up-to-date.
+  
+Start the migration tool by running the `dart migrate` command
+in the directory that contains the package's `pubspec.yaml` file:
+
+```terminal
+$ dart migrate
+```
+
+If your package is ready to migrate,
+then the tool produces a line like the following:
+
+```terminal
+View the migration suggestions by visiting:
+
+  http://127.0.0.1:60278/Users/you/project/mypkg.console-simple?authToken=Xfz0jvpyeMI%3D
+```
+
+Visit that URL in a Chrome browser
+to see an interactive UI
+where you can guide the migration process:
+
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+
+For every variable and type annotation,
+you can see what nullability the tool infers.
+For example, in the preceding screenshot,
+the tool infers that the `ints` list (previously a list of `int`)
+in line 1 is nullable, and thus should be a list of `int?`.
+
+
+#### Understanding migration results
+
+To see the reasons for each change (or non-change),
+click its line number in the **Proposed Edits** pane.
+The reasons appear in the **Edit Details** pane.
+
+For example, consider the following code,
+from before null safety:
+
+```dart
+var ints = const <int>[0, null];
+var zero = ints[0];
+var one = zero + 1;
+var zeroOne = <int>[zero, one];
+```
+
+The default migration is backward compatible but not ideal:
```

Comment:
I don't actually see this migration if I run the tool on the code above.  The suggestion I see is:

```
 var ints = const <int?>[0, null];
  var zero = ints[0]!;
  var one = zero + 1;
  var zeroOne = <int >[zero, one];
```

Reply:
Ah, the pleasures of documenting tools that are still changing. Both what you see and what I saw are different from what's in https://github.com/dart-lang/sdk/tree/master/pkg/nnbd_migration#interactive-feedback-to-the-tool (even before I changed the code to make it more standard).

Reply:
Actually... even after updating I get these suggestions. Are you sure you don't with the latest flutter master release?

Reply:
If you want the build that is going out tomorrow, then you should use latest Flutter **dev** channel.

But I get the same results in latest master (Dart `2.12.0-50.0.dev`) and dev (Dart `2.12.0-29.10.beta`):
<img width="740" alt="Screenshot 2020-11-18 at 23 36 24" src="https://user-images.githubusercontent.com/13644170/99596631-f12ec580-29f6-11eb-90b0-6f7815a58953.png">

Reply:
I wonder if Leaf's putting this code in a function. I have it outside (top level).

Reply:
Confirmed that I get the same results as Leaf if I move the code into a function.